pkg/threshcrypto/tbls.go

```diff
@@ -0,0 +1,330 @@
+package threshcrypto
+
+import (
+	"crypto/cipher"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/drand/kyber"
+	bls12381 "github.com/drand/kyber-bls12381"
+	"github.com/drand/kyber/pairing"
+	"github.com/drand/kyber/share"
+	"github.com/drand/kyber/sign"
+	"github.com/drand/kyber/sign/tbls"
+	"golang.org/x/exp/slices"
+
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+// TBLSInst an instance of a BLS-based (t, len(members))-threshold signature scheme
+// It is capable of creating signature shares with its (single) private key share,
+// and validating/recovering signatures involving all group members.
+type TBLSInst struct {
+	t         int
+	members   []t.NodeID
+	scheme    sign.ThresholdScheme
+	sigGroup  kyber.Group
+	privShare *share.PriShare
+	public    *share.PubPoly
+}
+
+// Constructs a TBLS scheme using the BLS12-381 pairing, with signatures being points on curve G1,
+// and keys points on curve G2.
+func tbls12381Scheme() (pairing.Suite, sign.ThresholdScheme, kyber.Group, kyber.Group) {
+	suite := bls12381.NewBLS12381Suite()
+	scheme := tbls.NewThresholdSchemeOnG1(suite)
+	sigGroup := suite.G1()
+	keyGroup := suite.G2()
+
+	return suite, scheme, sigGroup, keyGroup
+}
+
+// TBLS12381Keygen constructs a set TBLSInst for a given set of member nodes and threshold T
+// using the BLS12-381 pairing, with signatures being points on curve G1,
+// and keys points on curve G2.
+func TBLS12381Keygen(T int, members []t.NodeID, randSource cipher.Stream) ([]*TBLSInst, error) {
+	N := len(members)
+	if !(T >= (N+1)/2 && N > 0) {
+		return nil, fmt.Errorf("TBLS requires t >= (n+1)/2 and n > 0, where n is the number of group members")
+	}
```

Comment:
Could you give a reference for this? I believe many applications need (t+1, 3t+1) threshold signatures.

Reply:
Ah oops, I had a "TODO" comment precisely to remember to review this check carefully, and must have accidentally deleted it at some point

Reply:
Ok, I reviewed the [literature](https://link.springer.com/content/pdf/10.1007/3-540-36288-6_3.pdf), and comitted a correct version, that supports the (t+1, 3t+1) use case.

It's hard (impossible?) to dissociate the check from the number of malicious nodes.

Reply:
I would suggest removing this check altogether as it is rather trivial and concerns rather the protocol using the threshold signatures than the threshold signatures themselves.

I.e., you can have a (2, n)-threshold signature scheme, why not. It's just that, in the classical BFT model, you are unlikely to find such a signature useful, but it is not a concern of the threshold signature module.

Reply:
Or am I missing something?
Where does `nByz < (N+1)/2` come from?

Reply:
My speculation is that this is the requirement not for the threshold signature scheme itself, but for the interactive signing protocol. However, I need to check.

Reply:
Sorry, for taking so much time. I believe that the `t < n/2` part in the paper comes from the way the threshold signature scheme is used (in particular, the DKG part and the protocol for signing a message), which is kind of irrelevant.
So, I stick to my suggestion to remove this check completely.